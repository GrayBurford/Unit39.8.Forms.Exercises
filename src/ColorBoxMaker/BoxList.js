import React, {useState} from 'react';
import './BoxList.css';
import Box from './Box';
import NewBoxForm from './NewBoxForm';
import { v4 as uuid } from 'uuid';


// Box state goes here
// Renders all box components and NewBoxForm component
function BoxList () {
    const INITIAL_STATE = [];
    const [boxes, setBoxes] = useState(INITIAL_STATE)

    function addBox (newBox) {
        setBoxes(boxes => [...boxes, {...newBox, id : uuid() }])
    }

    function handleDelete (evt) {
        evt.target.parentElement.remove();
    }

    return (
        <div className="BoxList">
            <h1>BoxList App</h1>

            <NewBoxForm addBox={addBox} />

            {boxes.map(({backgroundColor, width, height, id}) => 
                <Box 
                    handleDelete={handleDelete}
                    backgroundColor={backgroundColor} 
                    height={height} 
                    width={width} 
                    key={id}
                    id={id}
                />
            )}
        </div>
    )
}

export default BoxList;
