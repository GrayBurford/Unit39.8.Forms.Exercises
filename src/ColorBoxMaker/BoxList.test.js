import {render, fireEvent} from '@testing-library/react';
import BoxList from './BoxList';

function addBox (boxList, height = "100", width = "150", color = "pink") {
    const heightInput = boxList.getByLabelText("Height");
    const widthInput = boxList.getByLabelText("Width");
    const backgroundInput = boxList.getByLabelText("Background Color");
    fireEvent.change(backgroundInput, { target: { value: color } });
    fireEvent.change(widthInput, { target: { value: width } });
    fireEvent.change(heightInput, { target: { value: height } });
    const button = boxList.getByText("Add Box");
    fireEvent.click(button);
  }

it('renders BoxList component without crashing', () => {
    render(<BoxList />);
})

it('matches snapshot', () => {
    const { asFragment } = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
})

it("should add a new box successfully", function() {
    const boxList = render(<BoxList />);
  
    // no boxes yet
    expect(boxList.queryByText("X")).not.toBeInTheDocument();
  
    addBox(boxList);
  
    // expect to see a box
    const removeButton = boxList.getByText("X");
    expect(removeButton).toBeInTheDocument();
    // expect(removeButton.previousSibling).toHaveStyle(`
    //   background-color: pink;
    // `);
  
    // expect(asFragment()).toMatchSnapshot();
  });
  
  it("can remove a box", function() {
    const boxList = render(<BoxList />);
    addBox(boxList);
  
    const removeButton = boxList.getByText("X");
  
    // click the remove button and the box should be gone
    fireEvent.click(removeButton);
    expect(removeButton).not.toBeInTheDocument();
  });
  