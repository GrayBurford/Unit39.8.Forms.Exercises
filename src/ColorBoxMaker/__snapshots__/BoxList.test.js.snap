// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="BoxList"
  >
    <h1>
      BoxList App
    </h1>
    <form
      class="NewBoxForm"
    >
      <div>
        <label
          for="backgroundColor"
        >
          Background Color
        </label>
        <select
          id="backgroundColor"
          name="backgroundColor"
        >
          <option
            value=""
          >
            -- choose a color --
          </option>
          <option
            value="black"
          >
            Black
          </option>
          <option
            value="pink"
          >
            Pink
          </option>
          <option
            value="blue"
          >
            Blue
          </option>
          <option
            value="purple"
          >
            Purple
          </option>
        </select>
      </div>
      <div>
        <label
          for="height"
        >
          Height
        </label>
        <input
          id="height"
          max="300"
          min="50"
          name="height"
          placeholder="Box Height"
          step="25"
          type="range"
          value="50"
        />
         (50)
      </div>
      <div>
        <label
          for="width"
        >
          Width
        </label>
        <input
          id="width"
          max="300"
          min="50"
          name="width"
          placeholder="Box Width"
          step="25"
          type="range"
          value="50"
        />
         (50)
      </div>
      <button>
        Add Box
      </button>
    </form>
  </div>
</DocumentFragment>
`;
